backend/src/utils/userRecommendations.js

```diff
@@ -0,0 +1,213 @@
+const { NONSHAREABLE, TWOBEDROOM, HABITS, NEUTRAL, POSINF, NEGINF } = require('./constants');
+
+/**
+ * Rank a set of users based on their corresponding aggregate score.
+ *
+ * @param {Array<Array<Object>>} userScores - [userId, score] pair
+ * @returns {Array<ObjectID>} - user IDs ranked based on their score
+ */
+const generateUserRecommendations = (userScores) => {
+    const userRanking = userScores.sort((user1, user2) => user2[1] - user1[1]);
+    return userRanking.map((user) => user[0]);
+};
+
+/**
+ * Calculate compatibility score for each possible match with the current user
+ *
+ * @param {User Object} currentUserPreferences - preferences of current user seeking roommates
+ * @param {User Object} potentialMatchesPreferences - preferences of potential roommates
+ * @returns {Array<Object>} - list of user id and score pairs
+ */
+const generateUserScores = (currentUserPreferences, potentialMatchesPreferences) => {
+    let userScores = [];
+    const minMaxRanges = calculateMinMaxRanges(currentUserPreferences, potentialMatchesPreferences);
+    potentialMatchesPreferences.forEach((matchPreferences) => {
+        let currentMatchScore = 0;
+
+        const categoricalMetrics = [aggregateCategorialPreferenceScores, petFriendlinessScore, housingTypeScore];
+        const numericalMetrics = [roommateCountScore, leaseLengthScore, priceRangeScore];
+
+        categoricalMetrics.forEach((metric) => {
+            currentMatchScore += metric(currentUserPreferences, matchPreferences);
+        });
+
+        numericalMetrics.forEach((metric) => {
+            currentMatchScore += metric(currentUserPreferences, matchPreferences, minMaxRanges);
+        });
+
+        userScores.push([matchPreferences.userId, currentMatchScore]);
+    });
+    return userScores;
+};
+
+/**
+ * This helper will aggrerate scores for the following preferences of a user:
+ * [smoking, partying, drinking, noise]
+ *
+ * @param {User Object} currentUser - preferences of current user seeking roommates
+ * @param {User Object} possibleMatch - preferences of potential roommates
+ * @returns {Number} - matching score
+ */
+const aggregateCategorialPreferenceScores = (currentUser, possibleMatch) => {
+    let score = 0;
+
+    HABITS.forEach((preference) => {
+        if (currentUser[preference] === possibleMatch[preference]) {
+            score += 1;
+        } else if (currentUser[preference] === NEUTRAL || possibleMatch[preference] === NEUTRAL) {
+            score += 0.5;
+        }
+    });
+
+    return score;
+};
+
+/**
+ * This helper will aggregrate whether the users have the same pet preferences or not
+ *
+ * @param {User Object} currentUser - preferences of current user seeking roommates
+ * @param {User Object} possibleMatch - preferences of potential roommates
+ * @returns {Number} - matching score
+ */
+const petFriendlinessScore = (currentUser, possibleMatch) => {
+    return currentUser.petFriendly === possibleMatch.petFriendly;
```

Comment:
Technically, this returns a Boolean, not Number.

Reply:
Valid, implicit type casting not good -_-